cpu/rpx0xx/periph/timer.c

```diff
@@ -0,0 +1,241 @@
+}
+
+static inline void _timer_reset(tim_t dev)
+{
+    DEV(dev)->TIMELW = 0; /* always write timelw before timehw */
+    DEV(dev)->TIMEHW = 0; /* writes do not get copied to time until timehw is written */
+}
+
+static inline void _timer_enable_periodic(tim_t dev, int channel, uint8_t flags)
+{
+    _timer_flag_periodic[dev] |= (1U << channel);
+    if (flags & TIM_FLAG_RESET_ON_MATCH) {
+        _timer_flag_reset[dev] |= (1U << channel);
+    }
+    else {
+        _timer_flag_reset[dev] &= ~(1U << channel);
+    }
+}
+
+static inline void _timer_disable_periodic(tim_t dev, int channel)
+{
+    _timer_flag_periodic[dev] &= ~(1U << channel);
+}
+
+static inline bool _timer_is_periodic(tim_t dev, int channel)
+{
+    return !!(_timer_flag_periodic[dev] & (1U << channel));
+}
+
+static inline bool _timer_reset_on_match(tim_t dev, int channel)
+{
+    return !!(_timer_flag_reset[dev] & (1U << channel));
+}
+
+static inline void _irq_enable(tim_t dev)
+{
+    for (uint8_t i = 0; i < timer_config[dev].chn; i++) {
+        NVIC_EnableIRQ(timer_config[dev].ch[i].irqn);
+        io_reg_atomic_set(&DEV(dev)->INTE.reg, (1U << i));
+    }
+}
+
+static void _isr(tim_t dev, int channel)
+{
+    /* clear latched interrupt */
+    io_reg_atomic_clear(&DEV(dev)->INTR.reg, 1U << channel);
+
+    if (_timer_ctx_cb[dev]) {
+        _timer_ctx_cb[dev](_timer_ctx_arg[dev], channel);
+    }
+
+    if (_timer_is_periodic(dev, channel)) {
+        if (_timer_reset_on_match(dev, channel)) {
+            _timer_reset(dev);
+        }
+        /* rearm */
+        *ALARM(dev, channel) = *ALARM(dev, channel);
+    }
+    cortexm_isr_end();
+}
+
+int timer_init(tim_t dev, uint32_t freq, timer_cb_t cb, void *arg)
+{
+    if (dev >= TIMER_NUMOF) {
+        return -ENODEV;
+    }
+    /* The timer must run at 1000000 Hz (µs precision)
+       because the number of cycles per µs is shared with the watchdog.
+       The reference clock (clk_ref) is devided by WATCHDOG->TICK.bits.CYCLES
+       to generate µs ticks.
+     */
+    assert(freq == US_PER_SEC); (void)freq;
+    _timer_ctx_cb[dev] = cb;
+    _timer_ctx_arg[dev] = arg;
+    periph_reset(RESETS_RESET_timer_Msk);
+    periph_reset_done(RESETS_RESET_timer_Msk);
+    io_reg_write_dont_corrupt(&WATCHDOG->TICK.reg,
+                              (CLOCK_XOSC / MHZ(1)) << WATCHDOG_TICK_CYCLES_Pos,
+                              WATCHDOG_TICK_CYCLES_Msk);
+    _irq_enable(dev);
+    return 0;
+}
+
+int timer_set(tim_t dev, int channel, unsigned int timeout)
+{
+    if (dev >= TIMER_NUMOF) {
+        return -ENODEV;
+    }
+    if (channel < 0 || channel >= timer_config[dev].chn) {
+        return -EINVAL;
+    }
+    unsigned state = irq_disable();
+    _timer_disable_periodic(dev, channel);
+    /* an alarm interrupt matches on the lower 32 bit of the 64 bit timer counter */
+    uint64_t target = DEV(dev)->TIMERAWL + timeout;
+    *ALARM(dev, channel) = (uint32_t)target;
+    irq_restore(state);
```

Comment:
Couldn't it happen that on a value of `timeout = 0` this has to value for almost a full period, if the clock tick happens between line 147 and 148?

Reply:
I tried to force the timer interrupt to trigger if `timeout == 0`.
```C
if (!timeout) {
    io_reg_atomic_set(&DEV(dev)->INTF.reg, 1U << channel);
}
```
But that did not work. When the line executes gdb showed:
```
(gdb) 
target not halted
target rp2040.cpu was not halted when step was requested
CMSIS-DAP command mismatch. Expected 0x5 received 0x0
```
and the program freezed.

Is it ok to execute the callback immediately?
```C
    if (!timeout) {
        if (_timer_ctx_cb[dev]) {
            _timer_ctx_cb[dev](_timer_ctx_arg[dev], channel);
        }
    }
```